### ISSUE TITLE
Guzzle long-term support plans?

### ISSUE THREAD

**Issue by wimleers** [guzzle/guzzle#2358]
Drupal has used Guzzle for a number of years. We will be shipping Drupal 9 next year, and are hence evaluating all of our dependencies. Hence this issue :)

**Will Guzzle 6 be supported through at least the end of 2023 (which is when Drupal 9 will go EOL)?**

**Comment by GrahamCampbell** [comment]
Is Drupal built on Symfony components. Symfony has a new HTTP Client component which is becoming stable in Symfony v4.4.0 and v5.0.0, and will be supported for a long time. Maybe that would be a good replacement for Guzzle, with well defined maintenance and security periods?

**Comment by xjm** [comment]
Thanks @GrahamCampbell. Part of the reason for us reaching out is exactly that we need to decide whether to add a dependency on the HTTP client component, and whether we do that for Drupal 9 (released June 2020 and EOL probably in Dec 2023) or Drupal 10 (release schedule not yet totally defined but we might have to ship it in June 2022 or so because Drupal 9 is going to depend on Symfony 4.4 and therefore be tied to their end-of-life. Currently, `symfony/http-client` is missing some functionality we need that Guzzle provides.

If we know more about the Guzzle maintainers' longterm plans for Guzzle itself, that will help us plan for the future. We actively collaborate with Symfony maintainers and so the new component is interesting to us, but Guzzle has also historically been really great for us about security coverage for old minors and coordinated disclosure, and given our release schedule and the current state of `symfony/http-client` it would be less disruptive for us if we can depend on Guzzle 6 at least through Drupal 9's end of life. :) Right now we are just looking for insight into Guzzle's support future (especially security coverage).

Thanks so much!

**Comment by GrahamCampbell** [comment]
Right. Thanks for the details. :)

**Comment by sagikazarmark** [comment]
Our long-term plan for Guzzle (probably 7) is adding PSR-18 support and dropping PHP 5.x support. Actually PSR-18 requires PHP 7, so it depends on whether we decide to drop PHP 5.x support in Guzzle 6 or 7. There are pro-contra arguments on both sides.

Other than that there isn't much that we plan to change.

In case we decide to drop PHP 5.x in Guzzle 6 (BTW: would that be a problem for you?), it will remain the main version. If not, Guzzle 5 will become EOL and Guzzle 6 and 7 will be the supported versions. Since the code will essentially be the same, we will support both versions.

**Comment by xjm** [comment]
This is exactly the information we needed; thank you so much for the reply!

We've dropped support for PHP 5 already (the last release that both supported it and used Guzzle will be EOL in 3 months). So Drupal at least is okay if you make a chance to the PHP version requirements of Guzzle 6 to be 7.0+, and we can continue to depend on it for a few more years.

As an aside, something to consider. We heard a lot of concerns when we increased our PHP version requirements in a minor release, folks saying it was too disruptive and against semver philosophy. A third path might be:

Guzzle 7: Exactly like Guzzle 6, just requiring PHP 7+
Guzzle 8: PSR-18 and any other BC-breakable things 

Like I said, this won't affect Drupal one way or the other just so long as Guzzle 6 has BC other than the PHP version change and security coverage through (in our case) December 2023. :)

Thanks again. This is great news for us.

**Comment by xjm** [comment]
@sagikazarmark Yep agreed, thanks so much. We also did the PHP major version bump with the same reasoning, but we got pushback about it because even though we were following the semver spec, there is a really widespread expectation that semver means only doing major dependency updates in major versions. :) And thanks for the clarification about PSR-18; I remember wondering about the PHP 7 requirement when I looked at that PR awhile back.

**Comment by sagikazarmark** [comment]
I think the root cause is the fact that PHP 5.x to 7.x upgrade is not necessarily easy, especially in ...hm, "traditional" web development environments (aka. the sysadmin installed PHP with apt).

With the spread of container technologies where upgrading a PHP version is really just a push of a button this will probably get better. But we are not there yet. 😕 

PSR-18 requiring PHP 7 was also a decision for the crowds, at least for the few loud people shouting on the FIG mailgroup. Mainly because the exception implements `Throwable` and tools like PHPStan don't complain. Other than that (and the return type) it has no advantage at all.

**Comment by Nyholm** [comment]
TLDR: 
> Will Guzzle 6 be supported through at least the end of 2023 (which is when Drupal 9 will go EOL)?

No, we cannot make this promise. 

> Will Guzzle 7 be released soon?

Yes. Hopfully this year

> What should Drupal and other projects do?

Depend on an interface, not an implementation. Use PSR-18.
